src/gui/guiHyperText.cpp

```diff
@@ -0,0 +1,1069 @@
+			if (text[cursor] == L'\n')
+				cursor++;
+			// If text has begun, don't skip empty line
+			if (m_paragraph) {
+				endParagraph();
+				enterElement(ELEMENT_SEPARATOR);
+			}
+			escape = false;
+			continue;
+		}
+
+		if (c == L'\n') { // Unix breaks
+			// If text has begun, don't skip empty line
+			if (m_paragraph) {
+				endParagraph();
+				enterElement(ELEMENT_SEPARATOR);
+			}
+			escape = false;
+			continue;
+		}
+
+		if (escape) {
+			escape = false;
+			pushChar(c);
+			continue;
+		}
+
+		if (c == L'\\') {
+			escape = true;
+			continue;
+		}
+
+		// Tag check
+		if (c == L'<') {
+			u32 newcursor = parseTag(text, cursor);
+			if (newcursor > 0) {
+				cursor = newcursor;
+				continue;
+			}
+		}
+
+		// Default behavior
+		pushChar(c);
+	}
+
+	endParagraph();
+}
+
+void ParsedText::endElement()
+{
+	m_element = NULL;
+}
+
+void ParsedText::endParagraph()
+{
+	if (!m_paragraph)
+		return;
+
+	endElement();
+	m_paragraph = NULL;
+}
+
+void ParsedText::enterParagraph()
+{
+	if (!m_paragraph) {
+		m_paragraphs.emplace_back();
+		m_paragraph = &m_paragraphs.back();
+		m_paragraph->setStyle(m_style);
+	}
+}
+
+void ParsedText::enterElement(ElementType type)
+{
+	enterParagraph();
+
+	if (!m_element || m_element->type != type) {
+		m_paragraph->elements.emplace_back();
+		m_element = &m_paragraph->elements.back();
+		m_element->type = type;
+		m_element->tags = m_active_tags;
+		m_element->setStyle(m_style);
+	}
+}
+
+void ParsedText::pushChar(wchar_t c)
+{
+	// New word if needed
+	if (c == L' ' || c == L'\t')
+		enterElement(ELEMENT_SEPARATOR);
+	else
+		enterElement(ELEMENT_TEXT);
+
+	m_element->text += c;
+}
+
+ParsedText::Tag *ParsedText::newTag(const std::string &name, const AttrsList &attrs)
+{
+	endElement();
+	Tag *newtag = new Tag();
+	newtag->name = name;
+	newtag->attrs = attrs;
+	m_tags.push_back(newtag);
+	return newtag;
+}
+
+ParsedText::Tag *ParsedText::openTag(const std::string &name, const AttrsList &attrs)
+{
+	Tag *newtag = newTag(name, attrs);
+	m_active_tags.push_front(newtag);
+	return newtag;
+}
+
+bool ParsedText::closeTag(const std::string &name)
+{
+	bool found = false;
+	for (auto id = m_active_tags.begin(); id != m_active_tags.end(); ++id)
+		if ((*id)->name == name) {
+			m_active_tags.erase(id);
+			found = true;
+			break;
+		}
+	return found;
+}
+
+void ParsedText::globalTag(const AttrsList &attrs)
+{
+	for (auto const &attr : attrs) {
+		// Only page level style
+		if (attr.first == "margin" && check_integer(attr.second))
+			margin = strtol(attr.second.c_str(), NULL, 10);
+
+		if (attr.first == "valign") {
+			if (attr.second == "top")
+				valign = ParsedText::VALIGN_TOP;
+			if (attr.second == "bottom")
+				valign = ParsedText::VALIGN_BOTTOM;
+			if (attr.second == "middle")
+				valign = ParsedText::VALIGN_MIDDLE;
+		}
+
+		if (attr.first == "background") {
+			irr::video::SColor color;
+			if (attr.second == "none")
+				background_type = BACKGROUND_NONE;
```

Comment:
First code block also needs braces. 
https://www.kernel.org/doc/html/v5.0/process/coding-style.html#placing-braces-and-spaces